### PR TITLE
feat: add cargo package manager and refactor existing package

### LINKED ISSUE #143: Package: Support cargo package manager (Rust)
> [🐟](https://github.com/matchai/spacefish-bot) **spacefish-bot** — Pull request copied from: https://github.com/denysdovhan/spaceship-prompt/pull/617.

#### Description

Make package section support cargo package manager.

If there is a missing field `version` in Cargo.toml, the following error occur:
```
error: failed to parse manifest at `/path/to/Cargo.toml`

Caused by:
missing field `version` for key `package`
```
Currently, we use `cargo pkgid`.
```
$ cargo pkgid
> file:///path/to/tokio-rs/tokio#0.1.14
```
But it depends on `cargo.lock`. So we need handle the following error:
```
error: a Cargo.lock must exist for this command
```

__If you work on a rust package, you almost have a cargo.lock in package root directory except you just only clone a package without any work.__

#### Screenshot
image

### PR DESCRIPTION
**Pull request by tiffafoo**
## Description

Make package section support cargo package manager. Additionally, update current implementation of `__sf_section_package.fish` to use `jq` and `python` to go through `package.json` as `spaceship` does as a performance boost.

## Motivation and Context

Closes #143 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

- [X] I have tested using **MacOS**
- [ ] I have tested using **Linux**

## Checklist:

- [X] I have checked that no other PR duplicates mine
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [X] I have updated the tests accordingly.

### CODE REVIEW

File: functions/__sf_section_package.fish [hunk 1]
```diff
@@ -22,13 +23,39 @@ function __sf_section_package -d "Display the local package version"
...
-	type -q npm; or return
+	# Exit if there is no package.json or Cargo.toml
+	if not test -e ./package.json; and not test -e ./Cargo.toml
+		return
+	end
+
+	set -l package_version
+
+	# Check if package.json exists AND npm exists locally while supressing output to just exit code (-q)
+	if test -f ./package.json; and type -q npm
```
Comment (matchai) [inline functions/__sf_section_package.fish]: I would suggest doing the `type -q npm` check before the file check for a sooner return for non-npm users.
```suggestion
	if type -q npm; and test -f ./package.json
```

File: functions/__sf_section_package.fish [hunk 2]
```diff
@@ -22,13 +23,39 @@ function __sf_section_package -d "Display the local package version"
...
+		# Check if jq (json handler) exists locally. If yes, check in package.json version
+		if type -q jq
+			set package_version (jq -r '.version' package.json 2>/dev/null)
+		# Check if python exists locally, use json to check version in package.json
+		else if type -q python
+			set package_version (python -c "import json; print(json.load(open('package.json'))['version'])" 2>/dev/null)
+		# Check if node exists locally, use it to check version of package.json
+		else if type -q node
+			set package_version (node -p "require('./package.json').version" 2> /dev/null)
```
Comment (matchai) [inline functions/__sf_section_package.fish]: small nit for consistency:
```suggestion
			set package_version (node -p "require('./package.json').version" 2>/dev/null)
```

File: functions/__sf_section_package.fish [hunk 3]
```diff
@@ -22,13 +23,39 @@ function __sf_section_package -d "Display the local package version"
...
+		end
+	end
+
+	# Check if Cargo.toml exists and cargo command exists
+	# and use cargo pkgid to figure out the package
+	if test -f ./Cargo.toml; and type -q cargo
```
Comment (matchai) [inline functions/__sf_section_package.fish]: Same as above:
```suggestion
	if type -q cargo; and test -f ./Cargo.toml
```

**Review by matchai — DISMISSED** [review]
Looks great @sirMerr! 😄 
Just a couple small suggested changes and we're ready to merge 🏎💨

**Comment by matchai** [comment]
@sirMerr You may want to update the documentation in Options.md to mention the new Cargo support in the package section: https://github.com/matchai/spacefish/blob/master/docs/Options.md#package-version-package

Here is the equivalent spaceship documentation: https://github.com/denysdovhan/spaceship-prompt/blob/master/docs/Options.md#package-version-package

File: tests/__sf_section_package.test.fish [hunk 4]
```diff
@@ -0,0 +1,72 @@
+source $DIRNAME/spacefish_test_setup.fish
+
+function setup
+	spacefish_test_setup
+	mock cargo pkgid 0 "echo \"file:///Users/sirMerr/Development/test-rust#0.1.0\""
```
Comment (Snuggle) [inline tests/__sf_section_package.test.fish:5]: What's going on here? :thinking: 
`\"file:///Users/sirMerr/Development/test-rust#0.1.0\""`

File: functions/__sf_section_package.fish [hunk 5]
```diff
@@ -22,13 +23,39 @@ function __sf_section_package -d "Display the local package version"
...
+		# Handle missing field `version` in Cargo.toml.
+		# `cargo pkgid` needs Cargo.lock to exists too. If
+		# it doesn't, do not show package version
+		set -l pkgid (cargo pkgid 2>&1)
+		# Early return on error
+		echo $pkgid | grep -q "error:"; and return
 
-	set -l version_line (grep -E '"version": "v?([0-9]+\.){1,}' package.json)
-	set -l package_version (string split \" $version_line)[4]
+		set package_version (string match -r '#(.*)' $pkgid)[2]
```
Comment (Snuggle) [inline functions/__sf_section_package.fish:58]: Could we please get a comment here to explain what it's supposed to do? :slightly_smiling_face:

**Review by Snuggle — COMMENTED** [review]
I think a couple things here need a little more explanation, but I can't see anything glaring. :slightly_smiling_face:

File: tests/__sf_section_package.test.fish (hunk 4 above)
Comment (tiffafoo) [reply tests/__sf_section_package.test.fish:5]: screen shot 2019-03-07 at 8 50 35 pm

When you run `cargo pkgid`, this the output ^^. I'm creating a mock for when we run that [here](https://github.com/matchai/spacefish/pull/171/files/868664b7c3a449db1f61b4ce6fa2876fd214840e#diff-51d50e9fcd7a849b18a3e1a216350dafR53)

**Review by matchai — APPROVED** [review]
Looks great! 😄 
Thanks a lot, Tiff!

🚀🚀🚀🚀

**Comment by matchai** [comment]
@all-contributors please add @sirMerr for tests

**Pull request merged**